meta-client/src/main/java/org/apache/helix/metaclient/impl/zk/ZkMetaClient.java

```diff
@@ -247,4 +249,56 @@ public boolean[] set(List keys, List values, List version) {
   public List<OpResult> transactionOP(Iterable iterable) {
     return null;
   }
+
+  /**
+   * A converter class to transform {@link DataChangeListener} to {@link IZkDataListener}
+   */
+  static class DataListenerConverter implements IZkDataListener {
+    private final DataChangeListener _listener;
+
+    DataListenerConverter(DataChangeListener listener) {
+      _listener = listener;
+    }
+
+    private DataChangeListener.ChangeType convertType(Watcher.Event.EventType eventType) {
+      switch (eventType) {
+        case NodeCreated: return DataChangeListener.ChangeType.ENTRY_CREATED;
+        case NodeDataChanged: return DataChangeListener.ChangeType.ENTRY_UPDATE;
+        case NodeDeleted: return DataChangeListener.ChangeType.ENTRY_DELETED;
+        default: throw new IllegalArgumentException("EventType " + eventType + " is not supported.");
+      }
+    }
+
+    @Override
+    public void handleDataChange(String dataPath, Object data) throws Exception {
+      throw new UnsupportedOperationException("handleDataChange(String dataPath, Object data) is not supported.");
+    }
+
+    @Override
+    public void handleDataDeleted(String dataPath) throws Exception {
+      handleDataChange(dataPath, null, Watcher.Event.EventType.NodeDeleted);
+    }
+
+    @Override
+    public void handleDataChange(String dataPath, Object data, Watcher.Event.EventType eventType) throws Exception {
+      _listener.handleDataChange(dataPath, data, convertType(eventType));
+    }
+
+    @Override
+    public boolean equals(Object o) {
```

Comment:
why do we need this?

Reply:
Because this is a converter class, as wrapper for the old listener interface. This is to make sure the JVM can recognize the underlying listener when we add/remove the listener.